api/client/client.go

```diff
@@ -116,15 +116,20 @@ func newClient(cfg Config, dialer ContextDialer, tlsConfig *tls.Config) *Client
 // no dialer is given, the first address will be used. This address must
 // be an auth server address.
 func connectInBackground(ctx context.Context, cfg Config) (*Client, error) {
-	tlsConfig, err := cfg.Credentials[0].TLSConfig()
+	var err error
+	params := connectParams{cfg: cfg}
+
+	params.tlsConfig, err = cfg.Credentials[0].TLSConfig()
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
 	if cfg.Dialer != nil {
-		return dialerConnect(ctx, cfg, cfg.Dialer, tlsConfig)
+		params.dialer = cfg.Dialer
```

Comment:
params already has cfg field set, so do you really need to fill in dialer and addr (below) as well?

Reply:
I made it so that the `cfg.Dialer` is pulled out of cfg automatically. However, with how `synconnect` interacts with addrs right now, the later is not really possible. However I think a lot of this client connection logic is still confusing so I'm going to try to refactor it a bit in a separate PR.